File: frontend/lib/schemas/v2/opLog.ts
/**
 * op-log schemas — Wave U-3 戻る/進む (undo/redo).
 *
 * GET /api/v1/schedule/v2/op-log/state → OpLogState
 * POST .../undo / .../redo → OpLogState (新しい state を含む)
 *
 * BE 並行実装中のため欠落フィールドは .catch() で寛容にデフォルト値へフォールバック。
 * レスポンス全体が欠落 / 不正でも { can_undo: false, can_redo: false, ... } を返す。
 */
import { z } from 'zod';

/**
 * GET /api/v1/schedule/v2/op-log/state および
 * POST .../undo / .../redo レスポンスの共通スキーマ。
 *
 * 寛容パース: フィールド欠落は安全なデフォルト (undo/redo 不可) にフォールバック。
 */
export const opLogStateSchema = z
  .object({
    can_undo: z.boolean().catch(false),
    can_redo: z.boolean().catch(false),
    undo_label: z.string().nullable().catch(null),
    redo_label: z.string().nullable().catch(null),
  })
  .catch({
    can_undo: false,
    can_redo: false,
    undo_label: null,
    redo_label: null,
  });

export type OpLogState = z.infer<typeof opLogStateSchema>;

File: backend/app/schemas/v2/op_log.py
"""操作ジャーナル (schedule_op_log) のリクエスト/レスポンス スキーマ (Wave U-3)."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


class OpLogStateResponse(BaseModel):
    """GET /v2/op-log/state レスポンス.

    自分のその週のスタック状態。D-4: 自分の操作のみ対象。
    """

    model_config = ConfigDict(extra="forbid")

    can_undo: bool
    can_redo: bool
    undo_label: str | None = None  # 次の undo 対象グループのラベル（cannot undo → None）
    redo_label: str | None = None  # 次の redo 対象グループのラベル（cannot redo → None）


class UndoRedoRequest(BaseModel):
    """POST /v2/op-log/undo and /v2/op-log/redo のリクエスト."""

    model_config = ConfigDict(extra="forbid")

    iso_year: int = Field(ge=2000, le=2100)
    iso_week: int = Field(ge=1, le=53)


class UndoRedoResponse(BaseModel):
    """POST /v2/op-log/undo and /v2/op-log/redo のレスポンス.

    実行後の state を含める（FE の再フェッチ削減）。
    """

    model_config = ConfigDict(extra="forbid")

    state: OpLogStateResponse
